File: ContravariantBindingResolver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Ninject.Components;
using Ninject.Infrastructure;
using Ninject.Planning.Bindings;
using Ninject.Planning.Bindings.Resolvers;

namespace Ninject.ContravariantBindingResolver
{
    public class ContravariantBindingResolver : NinjectComponent, IBindingResolver
    {
        /// <summary>
        /// Returns any bindings from the specified collection that match the specified service.
        /// </summary>
        public IEnumerable<IBinding> Resolve(Multimap<Type, IBinding> bindings, Type service)
        {
            if (!service.IsGenericType)
            {
                return Enumerable.Empty<IBinding>();
            }

            Type genericType = service.GetGenericTypeDefinition();
            Type[] genericArguments = genericType.GetGenericArguments();
            if (genericArguments.Length != 1 ||
                !genericArguments[0].GenericParameterAttributes.HasFlag(GenericParameterAttributes.Contravariant))
            {
                return Enumerable.Empty<IBinding>();
            }

            Type argument = service.GetGenericArguments().Single();
            return from kvp in bindings
                from value in kvp.Value
                let key = kvp.Key
                let genericArgument = key.GetGenericArguments().Single()
                where
                key.IsGenericType && key.GetGenericTypeDefinition() == genericType && genericArgument != argument &&
                genericArgument.IsAssignableFrom(argument)
                select value;
        }
    }
}

File: Ninject.ContravariantBindingResolver.FSharp/ContravariantBindingResolver.fs
namespace Ninject.ContravariantBindingResolver.FSharp

open System.Reflection
open Ninject.Components
open Ninject.Planning.Bindings.Resolvers

type ContravariantBindingResolver () =
    inherit NinjectComponent ()
    interface IBindingResolver with
        member __.Resolve (bindings, service) =
            if not service.IsGenericType
            then Seq.empty
            else
                let genericType = service.GetGenericTypeDefinition ()
                let genericArguments = genericType.GetGenericArguments ()
                if Array.length genericArguments = 1 || 
                    GenericParameterAttributes.Contravariant
                    |> genericArguments.[0].GenericParameterAttributes.HasFlag
                    |> not
                then Seq.empty
                else
                    let argument = service.GetGenericArguments () |> Array.exactlyOne
                    bindings
                    |> Seq.map (fun (KeyValue (k, v)) -> 
                        k, k.GetGenericArguments () |> Array.exactlyOne, v)
                    |> Seq.filter (fun (key, genericArgument, _) -> 
                        key.IsGenericType && 
                        key.GetGenericTypeDefinition () = genericType &&
                        genericArgument <> argument && 
                        genericArgument.IsAssignableFrom argument)
                    |> Seq.collect (fun (_, _, v) -> v)